### ISSUE TITLE
[`json`] Numbers in scientific notation are not encoded correctly

### ISSUE THREAD

**Issue by LukeSavefrogs** [LukeSavefrogs/jython21-backport#6]
## Problem description
On Jython 2.2 numbers expressed in scientific notation are always expressed in scientific notation when not converted.

## Example
**Jython 2.2**:
```python
# Expected result
>>> 3.14 
3.14
>>> float(3.14)
3.14
>>> str(3.14)
'3.14'

# Expected result
>>> 3.14e+5    
314000.0
>>> float(3.14e+5)
314000.0
>>> str(3.14e+5)
'314000.0'

# WRONG
>>> 3.14e+10       
3.14E10
>>> float(3.14e+10)
3.14E10
>>> str(3.14e+10)
'3.14E10'
```

> See how `3.14e+5` is converted correctly while `3.14e+10` is left as is. 

**CPython 3.9**:
```python
# Expected result
>>> 3.14 
3.14
>>> float(3.14)
3.14
>>> str(3.14)
'3.14'

# Expected result
>>> 3.14e+5    
314000.0
>>> float(3.14e+5)
314000.0
>>> str(3.14e+5)
'314000.0'

# Expected result
>>> 3.14e+10
31400000000.0
>>> float(3.14e+10)
31400000000.0
>>> str(3.14e+10)
'31400000000.0'
```

**Comment by LukeSavefrogs** [comment]
It seems like this behaviour on **CPython 3.9** is present for exponents **greater than 15**...

Jython 2.2:
```python
>>> for x in range(0,20): print("%2d: %s" % (x, str(float("3.14e+%d" % x))))
... 
 0: 3.14
 1: 31.4
 2: 314.0
 3: 3140.0
 4: 31400.0
 5: 314000.0
 6: 3140000.0
 7: 3.14E7
 8: 3.14E8
 9: 3.14E9
10: 3.14E10
11: 3.14E11
12: 3.14E12
13: 3.14E13
14: 3.14E14
15: 3.14E15
16: 3.14E16
17: 3.14E17
18: 3.14E18
19: 3.14E19
>>> 
```

CPython 3.9:
```python
>>> for x in range(0,20): print("%2d: %s" % (x, str(float("3.14e+%d" % x))))
...
 0: 3.14
 1: 31.4
 2: 314.0
 3: 3140.0
 4: 31400.0
 5: 314000.0
 6: 3140000.0
 7: 31400000.0
 8: 314000000.0
 9: 3140000000.0
10: 31400000000.0
11: 314000000000.0
12: 3140000000000.0
13: 31400000000000.0
14: 314000000000000.0
15: 3140000000000000.0
16: 3.14e+16
17: 3.14e+17
18: 3.14e+18
19: 3.14e+19
>>>
```

**Comment by LukeSavefrogs** [comment]
The original `json.dumps` method seems to encode floating point numbers using the S.N. with the same rules as above (>15):

```python
>>> json.dumps(1234567890123456.12)
'1234567890123456.0'
>>> json.dumps(12345678901234567.12)
'1.2345678901234568e+16'
```

**Comment by LukeSavefrogs** [comment]
Counld be useful doing something along the lines of this:

```python
def format_float(number, round_precision = 6):
    """Formats a float number to a string representation without trailing zeros.

    Args:
        number (float): The number to format.
        round_precision (int, optional): The number of digits to round the number to. Pass `-1` to disable rounding. Defaults to 6.

    Returns:
        str: The formatted number.
    """
    if round_precision < 0: 
        format_str = "%f"
    else:
        format_str = ("%." + str(round_precision) + "f")
    formatted_number = str(format_str % number)
    
    if not "." in formatted_number: 
        return formatted_number
    
    while formatted_number.endswith("0"):
        if formatted_number[-2] == ".": break
        formatted_number = formatted_number[:-1]
    
    return formatted_number
```

However, should be tested throughly, since it produces some discrepancies:

**CPython 3.9**:
```python
>>> str(135234567217345.126507986654)
'135234567217345.12'
>>> format_float(135234567217345.126507986654)
'135234567217345.125'
```

**Jython 2.2**:
```python
>>> str(135234567217345.126507986654)
'1.3523456721734512E14'
>>> format_float(135234567217345.126507986654)
'135234567217345.12'
```

**Comment by LukeSavefrogs** [comment]
## Solution
I fixed this "bug" (which turned out being a false positive) in commit e43aac9, where I changed the comparison from a fixed string (`'31400000000.0'`) to a string casting operation (`str(3.14e10)`), which will always produce a valid result for the current Python version.

## Rationale
I noticed how this "bug" only affects how the number is shown, and not how it is parsed. 

Look at the following example:

**CPython 3.9**:
```pycon
>>> import polyfills.json as json
>>> json.dumps(3.14e10)
'31400000000.0'
>>> str(3.14e10)            
'31400000000.0'
>>> float("3.14E10")
31400000000.0
>>> float("31400000000.0") 
31400000000.0
>>> 3.14e10
31400000000.0
```

**Jython 2.2**:
```pycon
>>> import polyfills.json as json
>>> json.dumps(3.14e10)
3.14E10
>>> str(3.14e10)
'3.14E10'
>>> float("3.14E10")
3.14E10
>>> float("31400000000.0")
3.14E10
>>> 3.14e10
3.14E10
```

As we can see at the end even if the number is shown differently it is correctly parsed as a floating number (and represents the **same number**).

## Result
Tests runs now are [ending successfully](https://github.com/LukeSavefrogs/jython21-backport/actions/runs/5997616528):

![image](https://github.com/LukeSavefrogs/jython21-backport/assets/33452387/c3120d10-fdbd-45f5-9fec-d2818c65d1ca)
